src/webgpu/api/validation/vertex_access.spec.ts

```diff
@@ -1,91 +1,483 @@
+   */
+  testEncoderFinish(
+    vertexBufferLayouts: GPUVertexBufferLayout[],
+    indexBufferMappings: SetIndexBufferParam[],
+    vertexBufferMappings: SetVertexBufferParam[],
+    drawCall: DrawCall | null,
+    isFinishSuccess: boolean
+  ) {
+    for (const setPipelineBeforeBuffer of [false, true]) {
+      for (const useBundle of [false, true]) {
+        this.doBufferSettingAndDraw(
+          vertexBufferLayouts,
+          indexBufferMappings,
+          vertexBufferMappings,
+          drawCall,
+          isFinishSuccess,
+          setPipelineBeforeBuffer,
+          useBundle
+        );
+      }
+    }
+  }
+}
+
+export const g = makeTestGroup(F);
+
+// Buffer usage list for testing index and vertex buffer
+const bufferUsageListForTest: number[] = [
+  GPUConst.BufferUsage.MAP_WRITE,
+  GPUConst.BufferUsage.INDEX,
+  GPUConst.BufferUsage.VERTEX,
+  GPUConst.BufferUsage.INDEX | GPUConst.BufferUsage.VERTEX,
+];
+
+g.test('set_buffer_usage_validation')
+  .desc(
+    `
+In this test, we test the usage validation within setIndexBuffer and setVertexBuffer.
+    - Validate that buffer usage mismatch is catched by set*Buffer
+    - Validate that calling set*Buffer before/after setPipeline is the same
+
+Related set*Buffer validation rules:
+    - Index buffer
+        - buffer.[[usage]] contains INDEX.
+    - Vertex buffer
+        - buffer.[[usage]] contains VERTEX.
+`
+  )
+  .params(u =>
+    u
+      .combine('indexFormat', ['uint16', 'uint32'] as GPUIndexFormat[])
+      // Test all usage case for all 3 buffers
+      .combine('indexBufferUsage', bufferUsageListForTest)
+      .combine('veretxBufferUsage', bufferUsageListForTest)
+      .combine('instanceBufferUsage', bufferUsageListForTest)
+  )
+  .fn(t => {
+    const p = t.params;
+
+    const usages = [p.indexBufferUsage, p.veretxBufferUsage, p.instanceBufferUsage];
+    const buffers: GPUBuffer[] = usages.map(usage =>
+      t.createBufferWithState('valid', {
+        size: 128,
+        usage,
+        mappedAtCreation: false,
+      })
+    );
+
+    const indexBuffer = buffers[0];
+    const instanceBuffer = buffers[1];
+    const vertexBuffer = buffers[2];
+
+    const indexBufferParams: SetIndexBufferParam[] = [
+      { indexFormat: p.indexFormat, buffer: indexBuffer },
+    ];
+    const vertexBufferParams: SetVertexBufferParam[] = [
+      { slot: 0, buffer: instanceBuffer },
+      { slot: 1, buffer: vertexBuffer },
+    ];
+
+    const isUsageValid =
+      (usages[0] & GPUBufferUsage.INDEX) === GPUBufferUsage.INDEX &&
+      (usages[1] & GPUBufferUsage.VERTEX) === GPUBufferUsage.VERTEX &&
+      (usages[2] & GPUBufferUsage.VERTEX) === GPUBufferUsage.VERTEX;
+
+    // Test setIndexBuffer and setVertexBuffer without calling draw
+    t.testEncoderFinish(
+      t.getBasicVertexBufferLayouts({}),
+      indexBufferParams,
+      vertexBufferParams,
+      null,
+      isUsageValid
+    );
+  });
+
+g.test('set_buffer_parameter_validation')
```

Comment:
The one special case that wasn't covered in `setVertexBuffer.spec.ts` and `setIndexBuffer.spec.ts` is bound size = 0, given or computed, this case lead to Dawn issue [1058](https://bugs.chromium.org/p/dawn/issues/detail?id=1058).

Reply:
Actually in setVertexBuffer.spec.ts the case of given bound size = 0 is tested, but only test that no validation error in setVertexBuffer. I found this issue by calling draw after setting buffer and expect it can only draw 0 vertex, but instead it can draw vertices reading the whole buffer size.